In [ ]:
!pip install recbole
!pip install ray

In [15]:
import pandas as pd
import numpy as np
import torch

import sys
import logging
from recbole.config import Config
from recbole.data import create_dataset, data_preparation
from recbole.model.sequential_recommender import SASRec, BERT4Rec
from recbole.trainer import Trainer
from recbole.utils import init_seed, init_logger
from recbole.quick_start import load_data_and_model
from recbole.data import Interaction

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


## Трансформирование датасета KION в формат Recbole

In [5]:
inter_df = pd.read_csv('interactions_processed.csv')

In [6]:
!mkdir "recbole_data"

In [7]:
filename = f"recbole_data/recbole_data.inter"

In [8]:
inter_df['last_watch_dt'] = pd.to_datetime(inter_df['last_watch_dt']).astype('int64') // 10**9
inter_df.drop("total_dur", axis=1, inplace=True)

inter_df.rename(columns={
    "user_id": "user_id:token", 
    "item_id": "item_id:token", 
    "last_watch_dt": "timestamp:float", 
    "watched_pct": "watched_pct:float",
}, inplace=True)

inter_df.head(3)

user_id:token  item_id:token  timestamp:float  watched_pct:float
0         176549           9506       1620691200                 72
1         699317           1659       1622246400                100
2         656683           7107       1620518400                  0

In [9]:
inter_df.to_csv(filename, sep="\t", index=False)

## SASRec

In [ ]:
parameter_dict = {
    "worker": 2,
    "device": "GPU",
    "data_path": "",

    "USER_ID_FIELD": "user_id",
    "ITEM_ID_FIELD": "item_id",
    "TIME_FIELD": "timestamp",
    "load_col" : {"inter": ["user_id", "item_id", "timestamp", "watched_pct"]},

    "val_interval": {"watched_pct": "[10,inf)"},
    "user_inter_num_interval": "[40,inf)",
    "item_inter_num_interval": "[40,inf)",

    "epoch": 100,
    
    "eval_args": {
        "group_by": "user",
        "order": "TO",
        "split": {"RS": [8, 1, 1]},
        "mode": "full",
    },
    "metrics": ["MAP", "Recall", "Precision", "MRR", "NDCG", "ItemCoverage", "GiniIndex"],
    "valid_metric": "MRR@10",

    "hidden_size": 128,
    "n_layers": 2,
    "n_heads": 4,
    "hidden_act": "swish",
    "loss_type": "BPR",
    "train_neg_sample_args": {
        "distribution": "popularity",
        "sample_num": 1,

    },
}

In [ ]:
# conf inittialization
config = Config(model='SASRec', dataset='recbole_data', config_dict=parameter_dict)

# init random seed
init_seed(config["seed"], config["reproducibility"])

# logger init
init_logger(config)

logger = logging.getLogger()
logFormatter = logging.Formatter("%(asctime)s [%(levelname)-5.5s]  %(message)s")
# Create handlers
c_handler = logging.StreamHandler(sys.stdout)
c_handler.setFormatter(logFormatter)
c_handler.setLevel("INFO")

f_handler = logging.FileHandler("logs.txt")
f_handler.setFormatter(logFormatter)
f_handler.setLevel("INFO")

logger.addHandler(c_handler)
logger.addHandler(f_handler)

In [ ]:
dataset = create_dataset(config)
dataset

recbole_data
The number of users: 6096
Average actions of users: 60.29975389663659
The number of items: 2157
Average actions of items: 170.46706864564007
The number of inters: 367527
The sparsity of the dataset: 97.20492062101417%
Remain Fields: ['user_id', 'item_id', 'timestamp', 'watched_pct']

In [ ]:
train_data, valid_data, test_data = data_preparation(config, dataset)

In [ ]:
results = dict()

sasrec = SASRec(config, train_data.dataset).to(config["device"])
trainer = Trainer(config, sasrec)

# training
sas_best_valid_score, sas_best_valid_results = trainer.fit(
    train_data,
    valid_data,
)

# evaluation
sas_test_result = trainer.evaluate(test_data)

results["SASRec"] = sas_test_result

In [ ]:
results

{'SASRec': OrderedDict([('map@10', 0.0152),
              ('recall@10', 0.0497),
              ('precision@10', 0.005),
              ('mrr@10', 0.0152),
              ('ndcg@10', 0.0232),
              ('itemcoverage@10', 0.9838),
              ('giniindex@10', 0.7688)])}

## BERT4Rec

In [ ]:
# conf inittialization
config = Config(model='BERT4Rec', dataset='recbole_data', config_dict=parameter_dict)

# init random seed
init_seed(config["seed"], config["reproducibility"])

# logger init
init_logger(config)

In [ ]:
dataset = create_dataset(config)
train_data, valid_data, test_data = data_preparation(config, dataset)

In [ ]:
bert4rec = BERT4Rec(config, train_data.dataset).to(config["device"])
trainer = Trainer(config, bert4rec)

# training
bert_best_valid_score, bert_best_valid_results = trainer.fit(
    train_data,
    valid_data, 
    show_progress=True,
)

# evaluation
bert_test_result = trainer.evaluate(test_data)

results["BERT4Rec"] = bert_test_result

Evaluate   : 100%|███████████████████████████| 9/9 [00:03<00:00,  2.31it/s, GPU RAM: 5.23 G/14.75 G]


In [ ]:
df = pd.DataFrame(
    data=[
        list(results["SASRec"].values()),
        list(results["BERT4Rec"].values()),
        list(sas_best_valid_results.values()), 
        list(bert_best_valid_results.values()),
    ],
    columns=list(results["SASRec"].keys()),
    index=["SASRec (test)", "BERT4Rec (test)", "SASRec (best valid)", "BERT4Rec (best valid)"]
)
df

map@10  recall@10  precision@10  mrr@10  ndcg@10  \
SASRec (test)          0.0152     0.0497        0.0050  0.0152   0.0232   
BERT4Rec (test)        0.0161     0.0499        0.0050  0.0161   0.0239   
SASRec (best valid)    0.0167     0.0543        0.0054  0.0167   0.0254   
BERT4Rec (best valid)  0.0194     0.0595        0.0059  0.0194   0.0286   

                       itemcoverage@10  giniindex@10  
SASRec (test)                   0.9838        0.7688  
BERT4Rec (test)                 0.6528        0.8559  
SASRec (best valid)             0.9893        0.7360  
BERT4Rec (best valid)           0.6750        0.8516

Имплементирую в сервис модель BERT4Rec.

## Предсказания BERT4Rec. Интеграция в сервис

In [11]:
config, bert4rec, dataset, train_data, valid_data, test_data = \
    load_data_and_model("BERT4Rec-May-16-2023_10-44-04.pth")

16 May 21:53    INFO  
General Hyper Parameters:
gpu_id = 0
use_gpu = True
seed = 2020
state = INFO
reproducibility = True
data_path = recbole_data
checkpoint_dir = saved
show_progress = True
save_dataset = False
dataset_save_path = None
save_dataloaders = False
dataloaders_save_path = None
log_wandb = False

Training Hyper Parameters:
epochs = 300
train_batch_size = 2048
learner = adam
learning_rate = 0.001
train_neg_sample_args = {'distribution': 'popularity', 'sample_num': 1, 'alpha': 1.0, 'dynamic': False, 'candidate_num': 0}
eval_step = 1
stopping_step = 10
clip_grad_norm = None
weight_decay = 0.0
loss_decimal_place = 4

Evaluation Hyper Parameters:
eval_args = {'group_by': 'user', 'order': 'TO', 'split': {'RS': [8, 1, 1]}, 'mode': 'full'}
repeatable = True
metrics = ['MAP', 'Recall', 'Precision', 'MRR', 'NDCG', 'ItemCoverage', 'GiniIndex']
topk = [10]
valid_metric = MRR@10
valid_metric_bigger = True
eval_batch_size = 4096
metric_decimal_place = 4

Dataset Hyper Parameters:
field_

In [158]:
dataset.field2id_token

{'user_id': array(['[PAD]', '176549', '699317', ..., '163369', '753860', '942675'],
       dtype='<U7'),
 'item_id': array(['[PAD]', '9506', '1659', ..., '11864', '5897', '13624'],
       dtype='<U5')}

Оставим последние 50 просмотров пользователя и сделем предсказание - 10 наиболее вероятных айтемов.

In [159]:
max_sequence_len = 50

train_df = pd.DataFrame(data={
    "user_id": dataset["user_id"],
    "item_id": dataset["item_id"],
    "timestamp": dataset["timestamp"],
}).sort_values(["user_id", "timestamp"], ignore_index=True)

train_df["order"] = train_df.groupby("user_id")["item_id"].cumcount(ascending=False) + 1
train_df = train_df[train_df.order <= max_sequence_len]
train_df.drop("order", axis=1, inplace=True)

train_df = train_df.groupby(by="user_id", as_index=False)["item_id"].agg(list)

train_df["item_length"] = train_df.item_id.apply(
    lambda items: 
    len(items) if len(items) <= max_sequence_len else max_sequence_len
)

train_df.rename(columns={"item_id": "item_id_list"}, inplace=True)
train_df

user_id                                       item_id_list  item_length
0           1  [2064, 2042, 648, 2, 1906, 806, 936, 1239, 192...           50
1           2  [289, 1661, 504, 828, 714, 943, 1, 809, 918, 1...           50
2           3  [231, 1279, 111, 1417, 1915, 1701, 486, 1920, ...           50
3           4  [443, 214, 445, 1668, 386, 1955, 240, 25, 366,...           50
4           5  [41, 1642, 734, 393, 938, 773, 162, 538, 221, ...           50
...       ...                                                ...          ...
6090     6091  [788, 2102, 166, 1060, 350, 1258, 1481, 1848, ...           42
6091     6092  [297, 824, 1051, 225, 677, 114, 240, 284, 16, ...           39
6092     6093  [748, 1042, 246, 1974, 567, 735, 336, 302, 114...           40
6093     6094  [938, 1686, 887, 516, 351, 485, 1904, 432, 131...           39
6094     6095  [1414, 707, 366, 146, 221, 1170, 468, 761, 173...           45

[6095 rows x 3 columns]

In [160]:
train_interaction = Interaction(train_df).to("cuda:0")

In [322]:
bert4rec.eval()

with torch.no_grad():
    scores = bert4rec.full_sort_predict(train_interaction)

scores = scores[:, 1:]
scores = scores.cpu().numpy()
scores.shape

(6095, 2156)

нулевой элемент во втором измерении scores[:, 0] - не является айтемом, 

так как всего уникальных айтемов = 2156

уникальных пользователей = 6095

Маппинги и чтение просмотренных айтемов:

In [310]:
import dill


user_id_to_token = dataset.field2id_token["user_id"]
user_id_to_uid = {int(user_id_to_token[uid]): uid for uid in range(1, len(user_id_to_token))}
uid_to_user_id = {v: k for k, v in user_id_to_uid.items()}

item_id_to_token = dataset.field2id_token["item_id"]
item_id_to_iid = {int(item_id_to_token[iid]): iid for iid in range(1, len(item_id_to_token))}
iid_to_item_id = {v: k for k, v in item_id_to_iid.items()}

with open("popular_in_category_model.dill", "rb") as f:
    user_id_to_watched_item_ids = dill.load(f)["user_to_watched_items_map"]
len(user_id_to_watched_item_ids)

962179

**Предсказания**:

In [367]:
user_id_to_bert4rec_recs = {}
k_recs = 10

for uid, u_scores in enumerate(scores):
    uid += 1
    user_id = uid_to_user_id[uid]
    
    watched_item_ids = user_id_to_watched_item_ids[user_id]

    n = k_recs + len(watched_item_ids)
    # iid уменьшенные на 1
    top_n_iids = np.flip(u_scores.argsort()[-n:], 0)
    
    answer = []
    for iid in top_n_iids:
        iid += 1
        if iid_to_item_id[iid] not in watched_item_ids:
            answer.append(iid_to_item_id[iid])
        if len(answer) == k_recs:
            break
    
    user_id_to_bert4rec_recs[user_id] = answer

len(user_id_to_bert4rec_recs)

6095

In [366]:
# оффлайн метод
import pickle


with open("user_id_to_bert4rec_recs.pickle", "wb") as f:
    pickle.dump(user_id_to_bert4rec_recs, f)

## Смотрим на рекомендации:

In [340]:
items_df = pd.read_csv("items_processed.csv")
interactions_df = pd.read_csv("interactions_processed.csv")

#### Пример 1.

Последние просмотренные 25 фильмов

In [356]:
test_user_id = int(np.random.choice(user_id_to_token[1:]))
test_item_ids = user_id_to_bert4rec_recs[test_user_id]

interactions_df[(interactions_df["user_id"] == test_user_id)] \
    .sort_values("last_watch_dt")[["user_id", "item_id", "last_watch_dt"]][-20:] \
    .merge(items_df[["item_id", "content_type", "title", "genres", "countries", "age_rating", "release_year_cat"]], on="item_id")

user_id  item_id last_watch_dt content_type  \
0    818325    14177    2021-08-11         film   
1    818325     2981    2021-08-14         film   
2    818325    16270    2021-08-14         film   
3    818325    14826    2021-08-14         film   
4    818325    13159    2021-08-14         film   
5    818325    11899    2021-08-15         film   
6    818325    11645    2021-08-16         film   
7    818325    12268    2021-08-16         film   
8    818325    13988    2021-08-16         film   
9    818325     9070    2021-08-16         film   
10   818325    13095    2021-08-16         film   
11   818325    13701    2021-08-16         film   
12   818325    10234    2021-08-16         film   
13   818325     4873    2021-08-16         film   
14   818325    14689    2021-08-16         film   
15   818325     6741    2021-08-16         film   
16   818325    10343    2021-08-18         film   
17   818325    16064    2021-08-18         film   
18   818325     9164    2021-08-21         film   
19   818325    12837    2021-08-21         film   

                             title  \
0          астерикс и тайное зелье   
1                академия вампиров   
2                       тайна коко   
3                         толстяки   
4                          рататуй   
5               вся правда о любви   
6                    и это пройдёт   
7                          маэстро   
8                           листок   
9                         рассказы   
10                  забыли пароль?   
11               я очень возбужден   
12                  прощай, робин!   
13                   последний акт   
14         джуманджи: зов джунглей   
15  мидо и музыкальные инструменты   
16           побег с планеты земля   
17             большое путешествие   
18                          валл-и   
19                      мадагаскар   

                                               genres       countries  \
0                                 мультфильм, фэнтези         франция   
1   детективы, драмы, триллеры, боевики, ужасы, фэ...             сша   
2                    мультфильм, фэнтези, приключения             сша   
3                                      драмы, комедии         испания   
4    мультфильм, приключения, драмы, фэнтези, комедии             сша   
5                                  мелодрамы, комедии  великобритания   
6                                   драмы, мультфильм         франция   
7                                 мультфильм, комедии         франция   
8                                   драмы, мультфильм           чехия   
9                                      драмы, комедии          россия   
10                                мультфильм, комедии         франция   
11                                            комедии         испания   
12                                  драмы, мультфильм         франция   
13                                мультфильм, комедии         франция   
14                          боевики, фэнтези, комедии             сша   
15                            мультфильм, музыкальные         франция   
16       фантастика, мультфильм, приключения, комедии     канада, сша   
17         семейное, мультфильм, приключения, комедии     россия, сша   
18                фантастика, мультфильм, приключения             сша   
19                   мультфильм, приключения, комедии             сша   

    age_rating release_year_cat  
0          6.0        2010-2020  
1         12.0        2010-2020  
2         12.0        2010-2020  
3         18.0        2000-2010  
4          0.0        2000-2010  
5         16.0        2000-2010  
6          0.0         2020_inf  
7          0.0        2010-2020  
8          0.0         2020_inf  
9         18.0        2010-2020  
10         0.0         2020_inf  
11        16.0        2010-2020  
12         0.0         2020_inf  
13         6.0        2010-2020  
14        16.0        2010-2020  
15         0.0         2020_inf  
16         0.0       

Будущие рекомендации

In [357]:
items_df[
    items_df.item_id.isin(test_item_ids)
][["content_type", "title", "genres", "countries", "age_rating", "release_year_cat"]]

content_type                            title  \
60            film                       бетховен 2   
4617          film                             джок   
7563          film  снежная королева 3. огонь и лед   
8033          film                       смурфики 2   
8601          film                    подводная эра   
9698          film                       зверополис   
13063         film                         смурфики   
13064         film                     мы – монстры   
13577         film                       гадкий я 2   
15886         film               гномео и джульетта   

                                                  genres  \
60                          мелодрамы, семейное, комедии   
4617           приключения, мультфильм, вестерн, комедии   
7563           мультфильм, фэнтези, приключения, комедии   
8033              семейное, мультфильм, фэнтези, комедии   
8601                    мультфильм, приключения, комедии   
9698         приключения, мультфильм, детективы, комедии   
13063          мультфильм, фэнтези, приключения, комедии   
13064                ужасы, мультфильм, фэнтези, комедии   
13577  мультфильм, приключения, фантастика, фэнтези, ...   
15886        мелодрамы, мультфильм, приключения, комедии   

                         countries  age_rating release_year_cat  
60                             сша         6.0        1990-2000  
4617                           сша         0.0        2010-2020  
7563                        россия         6.0        2010-2020  
8033                           сша         0.0        2010-2020  
8601                           сша         6.0        2010-2020  
9698                           сша         6.0        2010-2020  
13063                          сша         0.0        2010-2020  
13064               великобритания         6.0        2010-2020  
13577         сша, франция, япония         0.0        2010-2020  
15886  великобритания, канада, сша         0.0        2010-2020

#### Пример 2.

Последние просмотренные 25 фильмов

In [360]:
test_user_id = int(np.random.choice(user_id_to_token[1:]))
test_item_ids = user_id_to_bert4rec_recs[test_user_id]

interactions_df[(interactions_df["user_id"] == test_user_id)] \
    .sort_values("last_watch_dt")[["user_id", "item_id", "last_watch_dt"]][-20:] \
    .merge(items_df[["item_id", "content_type", "title", "genres", "countries", "age_rating", "release_year_cat"]], on="item_id")

user_id  item_id last_watch_dt content_type  \
0    664739     1873    2021-08-11         film   
1    664739    15051    2021-08-11         film   
2    664739     6267    2021-08-12         film   
3    664739    12741    2021-08-13       series   
4    664739      973    2021-08-13         film   
5    664739    15078    2021-08-15       series   
6    664739    12225    2021-08-15       series   
7    664739    16244    2021-08-15         film   
8    664739    11769    2021-08-16         film   
9    664739      468    2021-08-16       series   
10   664739    16023    2021-08-16       series   
11   664739    13851    2021-08-16         film   
12   664739    11790    2021-08-17         film   
13   664739     2954    2021-08-17         film   
14   664739    14809    2021-08-17         film   
15   664739     3978    2021-08-17         film   
16   664739     1287    2021-08-17         film   
17   664739    15108    2021-08-20         film   
18   664739     7851    2021-08-22         film   
19   664739     8160    2021-08-22       series   

                                  title  \
0                             секс трип   
1                        порочная связь   
2                            смурфики 2   
3                         шиммер и шайн   
4               фиксики. большой секрет   
5                       щенячий патруль   
6                               фиксики   
7                    предложение (2009)   
8                             мошенники   
9         азбука безопасности на дороге   
10                    сказочный патруль   
11                my little pony в кино   
12                             гадкий я   
13                              миньоны   
14                             красотка   
15               любовь с ограничениями   
16            терминатор: тёмные судьбы   
17                            жираф раф   
18                 быть астрид линдгрен   
19  английский алфавит с тётушкой совой   

                                               genres        countries  \
0                                    фэнтези, комедии              сша   
1                                               драмы  италия, франция   
2              семейное, мультфильм, фэнтези, комедии              сша   
3                  мультсериалы, фэнтези, приключения      канада, сша   
4       мультфильм, музыкальные, приключения, комедии           россия   
5           мультсериалы, спорт, приключения, комедии      канада, сша   
6                  мультсериалы, приключения, комедии           россия   
7                           драмы, мелодрамы, комедии              сша   
8                                  мелодрамы, комедии    германия, сша   
9                                        мультсериалы          украина   
10                                       мультсериалы           россия   
11  мультфильм, приключения, мюзиклы, фэнтези, ком...      канада, сша   
12                       мультфильм, фэнтези, комедии     сша, франция   
13       фантастика, мультфильм, приключения, комедии              сша   
14                                 мелодрамы, комедии              сша   
15                                 мелодрамы, комедии           россия   
16                   боевики, фантастика, приключения       китай, сша   
17                                           семейное       нидерланды   
18                                   драмы, биография    дания, швеция   
19                                       мультсериалы          украина   

    age_rating release_year_cat  
0         18.0        2010-2020  
1         18.0         2020_inf  
2          0.0        2010-2020  
3          0.0        2010-2020  
4          6.0        2010-2020  
5          0.0        2010-2020  
6          0.0        2010-2020  
7         16.0        2000-2010  
8         16.0        2000-2010  
9          0.0        2000-2010  
10         0.0        2010-2020  
11         6.0        2010-2020  
12         0.0        2010-2020

Будущие рекомендации

In [361]:
items_df[
    items_df.item_id.isin(test_item_ids)
][["content_type", "title", "genres", "countries", "age_rating", "release_year_cat"]]

content_type                 title  \
158           film                вперёд   
809           film             анастасия   
1554          film            тайна коко   
6990          film                 рио 2   
7939          film          кролик питер   
8698          film    алиса в зазеркалье   
9921          film                 моана   
10721         film  красавица и чудовище   
11254         film        суперсемейка 2   
13063         film              смурфики   

                                                  genres            countries  \
158    для детей, приключения, семейное, фэнтези, ком...                  сша   
809     мультфильм, приключения, драмы, мюзиклы, фэнтези                  сша   
1554                    мультфильм, фэнтези, приключения                  сша   
6990           мюзиклы, мультфильм, приключения, комедии                  сша   
7939           мультфильм, фэнтези, приключения, комедии                  сша   
8698                      семейное, фэнтези, приключения  великобритания, сша   
9921                        мультфильм, фэнтези, мюзиклы                  сша   
10721            мюзиклы, мультфильм, фэнтези, мелодрамы                  сша   
11254                фантастика, мультфильм, приключения                  сша   
13063          мультфильм, фэнтези, приключения, комедии                  сша   

       age_rating release_year_cat  
158           6.0         2020_inf  
809           6.0        1990-2000  
1554         12.0        2010-2020  
6990          0.0        2010-2020  
7939          6.0        2010-2020  
8698         12.0        2010-2020  
9921          6.0        2010-2020  
10721         6.0        1990-2000  
11254         6.0        2010-2020  
13063         0.0        2010-2020

#### Пример 3.

Последние просмотренные 25 фильмов

In [362]:
test_user_id = int(np.random.choice(user_id_to_token[1:]))
test_item_ids = user_id_to_bert4rec_recs[test_user_id]

interactions_df[(interactions_df["user_id"] == test_user_id)] \
    .sort_values("last_watch_dt")[["user_id", "item_id", "last_watch_dt"]][-20:] \
    .merge(items_df[["item_id", "content_type", "title", "genres", "countries", "age_rating", "release_year_cat"]], on="item_id")

user_id  item_id last_watch_dt content_type                     title  \
0    627400     3777    2021-08-17       series  драконы и всадники олуха   
1    627400    10876    2021-08-17         film                   гарфилд   
2    627400    10755    2021-08-17         film            время монстров   
3    627400    13243    2021-08-17         film               головоломка   
4    627400     3182    2021-08-19         film    ральф против интернета   
5    627400     8070    2021-08-19         film       приказ "уничтожить"   
6    627400     2025    2021-08-19         film             семейка крудс   
7    627400    14196    2021-08-19       series               пять невест   
8    627400     5754    2021-08-19         film            доктор стрэндж   
9    627400    14317    2021-08-20         film                     веном   
10   627400     9335    2021-08-20         film          подводная братва   
11   627400    10219    2021-08-21         film         железный рыцарь 2   
12   627400     3401    2021-08-21         film        зомби по имени шон   
13   627400    13290    2021-08-21         film           царь скорпионов   
14   627400      760    2021-08-21         film              легенда № 17   
15   627400       85    2021-08-22         film                     турбо   
16   627400    11985    2021-08-22         film         история игрушек 4   
17   627400     6467    2021-08-22         film               лови волну!   
18   627400     8482    2021-08-22         film    самолёты: огонь и вода   
19   627400     5681    2021-08-22         film                  самолеты   

                                               genres  \
0   мультсериалы, приключения, семейное, фэнтези, ...   
1                        мультфильм, фэнтези, комедии   
2                                      ужасы, фэнтези   
3                     фантастика, мультфильм, комедии   
4   мультфильм, приключения, фантастика, семейное,...   
5                               историческое, военные   
6                                 мультфильм, фэнтези   
7                                    военные, комедии   
8           боевики, фантастика, фэнтези, приключения   
9    популярное, фантастика, триллеры, боевики, ужасы   
10          мультфильм, фэнтези, приключения, комедии   
11  историческое, приключения, драмы, военные, бое...   
12                                     ужасы, комедии   
13                      боевики, фэнтези, приключения   
14                            драмы, спорт, биография   
15  мультфильм, приключения, спорт, фантастика, ко...   
16                       мультфильм, фэнтези, комедии   
17                         спорт, мультфильм, комедии   
18                   мультфильм, приключения, комедии   
19                   мультфильм, приключения, комедии   

                  countries  age_rating release_year_cat  
0                       сша         6.0        2010-2020  
1                       сша         0.0        2000-2010  
2                       сша        16.0        2010-2020  
3                       сша         6.0        2010-2020  
4                       сша         6.0        2010-2020  
5                    россия        12.0        2010-2020  
6                       сша         0.0        2010-2020  
7                    россия        16.0        2010-2020  
8                       сша        16.0        2010-2020  
9                       сша        16.0        2010-2020  
10                      сша        12.0        2000-2010  
11   великобритания, сербия        16.0        2010-2020  
12  великобритания, франция        16.0        2000-2010  
13   бельгия, германия, сша        12.0        2000-2010  
14                   россия         6.0        2010-2020  
15                      сша         0.0        2010-2020  
16                      сша         6.0        2010-2020  
17                      сша         0.0        2000-2010  
18                      сша         0.0        2010-2020  
19      

Будущие рекомендации

In [363]:
items_df[
    items_df.item_id.isin(test_item_ids)
][["content_type", "title", "genres", "countries", "age_rating", "release_year_cat"]]

content_type                                       title  \
3491          film             человек-паук: возвращение домой   
4188          film                                     тачки 3   
8754          film  ледниковый период 4: континентальный дрейф   
9003          film                                город героев   
9698          film                                  зверополис   
9848          film                                       ральф   
9921          film                                       моана   
10395         film                           ледниковый период   
10792         film                 человек-паук: вдали от дома   
15352         film                          холодное сердце ii   

                                             genres countries  age_rating  \
3491   боевики, популярное, фантастика, приключения       сша        16.0   
4188                            мультфильм, комедии       сша         6.0   
8754               мультфильм, приключения, комедии       сша         0.0   
9003       боевики, фантастика, мультфильм, комедии       сша         6.0   
9698    приключения, мультфильм, детективы, комедии       сша         6.0   
9848                   мультфильм, фэнтези, комедии       сша         6.0   
9921                   мультфильм, фэнтези, мюзиклы       сша         6.0   
10395     мультфильм, фэнтези, приключения, комедии       сша         0.0   
10792              боевики, фантастика, приключения       сша        12.0   
15352              фэнтези, мультфильм, музыкальные       сша         6.0   

      release_year_cat  
3491         2010-2020  
4188         2010-2020  
8754         2010-2020  
9003         2010-2020  
9698         2010-2020  
9848         2010-2020  
9921         2010-2020  
10395        2000-2010  
10792        2010-2020  
15352        2010-2020

#### Пример 4.

Последние просмотренные 25 фильмов

In [364]:
test_user_id = int(np.random.choice(user_id_to_token[1:]))
test_item_ids = user_id_to_bert4rec_recs[test_user_id]

interactions_df[(interactions_df["user_id"] == test_user_id)] \
    .sort_values("last_watch_dt")[["user_id", "item_id", "last_watch_dt"]][-20:] \
    .merge(items_df[["item_id", "content_type", "title", "genres", "countries", "age_rating", "release_year_cat"]], on="item_id")

user_id  item_id last_watch_dt content_type  \
0    944850    13865    2021-06-11         film   
1    944850    11778    2021-06-13         film   
2    944850    10152    2021-06-15         film   
3    944850     4930    2021-06-18         film   
4    944850     9059    2021-06-20         film   
5    944850    13223    2021-06-22         film   
6    944850     3999    2021-06-24         film   
7    944850    12356    2021-06-27         film   
8    944850    16444    2021-07-02         film   
9    944850    14998    2021-07-04         film   
10   944850    10634    2021-07-09         film   
11   944850    16442    2021-07-11         film   
12   944850    11929    2021-07-15         film   
13   944850     7389    2021-07-24         film   
14   944850    13701    2021-08-05         film   
15   944850    11268    2021-08-09         film   
16   944850     3031    2021-08-10         film   
17   944850     4905    2021-08-16         film   
18   944850     8584    2021-08-18         film   
19   944850    16484    2021-08-20         film   

                            title  \
0                        девятаев   
1               простые сложности   
2                          гамбит   
3                    муж двух жён   
4                      ловец снов   
5                      экстрасенс   
6                 джиперс криперс   
7                       13 грехов   
8              закатать в асфальт   
9                         невеста   
10           идеальный незнакомец   
11                  игра в правду   
12                       он и она   
13                ромовый дневник   
14              я очень возбужден   
15                         монгол   
16  техасская резня бензопилой 3d   
17                 братья систерс   
18                         пила 8   
19                   пила: начало   

                                               genres  \
0                         драмы, военные, приключения   
1                           драмы, мелодрамы, комедии   
2                                   криминал, комедии   
3                                             комедии   
4                                               ужасы   
5                              драмы, ужасы, триллеры   
6                                     ужасы, триллеры   
7                                     ужасы, триллеры   
8                  боевики, драмы, триллеры, криминал   
9                                     ужасы, триллеры   
10        криминал, детективы, драмы, триллеры, ужасы   
11                                            комедии   
12                                 мелодрамы, комедии   
13                        драмы, приключения, комедии   
14                                            комедии   
15  историческое, биография, приключения, драмы, в...   
16                                    ужасы, триллеры   
17                           драмы, криминал, вестерн   
18                         ужасы, криминал, детективы   
19                                    ужасы, триллеры   

                                  countries  age_rating release_year_cat  
0                                    россия        12.0         2020_inf  
1                               сша, япония        16.0        2000-2010  
2                                       сша        12.0        2010-2020  
3                            великобритания        16.0        2010-2020  
4                                       сша        18.0         2020_inf  
5                            великобритания        16.0        2010-2020  
6                             германия, сша        16.0        2000-2010  
7                                       сша        16.0        2010-2020  
8                               канада, сша        18.0        2010-2020  
9                                    россия        16.0        2010-2020  
10                                      сша        16.0        2000-2010  
11                                   россия        18.0 

Будущие рекомендации

In [365]:
items_df[
    items_df.item_id.isin(test_item_ids)
][["content_type", "title", "genres", "countries", "age_rating", "release_year_cat"]]

content_type              title  \
2819          film       цена страсти   
3158          film         американец   
3596          film    вне поля зрения   
4344          film              прочь   
4537          film       кривая линия   
7716          film  человек-невидимка   
8777          film       госфорд парк   
9255          film      мёрзлая земля   
10172         film        наблюдатель   
14160         film         предместье   

                                              genres  \
2819                      драмы, триллеры, мелодрамы   
3158                                 драмы, триллеры   
3596   криминал, драмы, триллеры, мелодрамы, комедии   
4344                      ужасы, триллеры, детективы   
4537                                 ужасы, триллеры   
7716          фантастика, ужасы, триллеры, детективы   
8777             драмы, криминал, детективы, комедии   
9255                                 драмы, триллеры   
10172           драмы, триллеры, криминал, детективы   
14160                             детективы, комедии   

                                    countries  age_rating release_year_cat  
2819                            германия, сша        16.0        2010-2020  
3158                      великобритания, сша        18.0        2010-2020  
3596                                      сша        16.0        1990-2000  
4344                              сша, япония        18.0        2010-2020  
4537                                      сша        18.0        2010-2020  
7716   австралия, великобритания, канада, сша        18.0         2020_inf  
8777              великобритания, италия, сша        16.0        2000-2010  
9255                                      сша        16.0        2010-2020  
10172                                     сша        18.0        2000-2010  
14160                                     сша        12.0        1980-1990